src/walker.ts

```diff
@@ -0,0 +1,62 @@
+import { PotentialStep, Pipeline, Conditional, Command, CommandStep } from '.';
+import { getAndCacheDependency } from './conditional';
+
+export type Entity = Pipeline | PotentialStep | Command;
+
+async function walkEntity<T extends Entity>(
+  entity: T,
+  nodeCache: Map<any, any>,
+  mutator?: (entity: Entity) => any,
+): Promise<T> {
+  if (nodeCache.has(entity)) {
+    return nodeCache.get(entity);
+  }
+
+  const cache = new Map();
+
+  if (entity instanceof Conditional) {
+    if (await entity.accept()) {
+      entity = <any>await getAndCacheDependency(cache, entity);
+    } else {
+      return entity;
+    }
+  }
+
+  if (entity instanceof Pipeline) {
+    const steps = entity.steps;
+    steps.forEach(async (step, idx) => {
+      const newStep = await walkEntity(step, nodeCache, mutator);
+      steps[idx] = newStep;
```

Comment:
does this work well with the other code we have? I don't think I ever intended overwriting the  steps array directly. It might work, it might not, unit tests will show :)

Reply:
It seems to work well via local testing. Will add unit tests to make sure though